rfcs/0077-layout-transform-padding.md

```diff
@@ -0,0 +1,3090 @@
+`BufferLoad` for which the predicate can be proven to be true.  If
+this simplified form is identical to the `else_case`, then the entire
+if/else block can be replaced with `then_case`.  Otherwise, this check
+is repeated to see if the `then_case` can be simplified down to the
+`else_case`.  If neither simplification holds, then no change is made.
+
+For example, consider the following example.  This is a 1-d
+convolution, where both the input and output buffers have a layout
+transformation applied.
+
+```python
+# Original function
+@T.prim_func
+def func(
+    A: T.Buffer[(16,), "float32"],
+    F: T.Buffer[(3,), "float32"],
+    B: T.Buffer[(14,), "float32"],
+):
+    with T.block('compute'):
+        for i in T.serial(14):
+            B[i] = 0.0
+            for f in T.serial(3):
+                B[i] = B[i] + A[i + f]
+
+
+# sched.transform_layout(block='compute', buffer='A', lambda i: [i//4, i%4])
+@T.prim_func
+def func(
+    A: T.Buffer[(4, 4), "float32"],
+    F: T.Buffer[(3,), "float32"],
+    B: T.Buffer[(14,), "float32"],
+):
+    with T.block('compute'):
+        for i in T.serial(14):
+            B[i] = 0.0
+            for f in T.serial(3):
+                B[i] = B[i] + A[(i + f) // 4, (i + f) % 4]
+
+
+# sched.transform_layout(block='compute', buffer='B', lambda i: [i//4, i%4], pad_value=0.0)
+@T.prim_func
+def func(
+    A: T.Buffer[(4, 4), "float32"],
+    F: T.Buffer[(3,), "float32"],
+    B: T.Buffer[(4, 4), "float32"],
+):
+    with T.block('compute'):
+        for i in T.serial(14):
+            B[i // 4, i % 4] = 0.0
+            for f in T.serial(3):
+                B[i // 4, i % 4] = B[i // 4, i % 4] + A[(i + f) // 4, (i + f) % 4]
+
+        for io,ii in T.grid(4,4):
+            if io==3 and ii>=2:
+                B[io,ii] = 0.0
+
+
+# sched.sequential_buffer_access(block='compute', buffer='B')
+@T.prim_func
+def func(
+    A: T.Buffer[(4, 4), "float32"],
+    F: T.Buffer[(3,), "float32"],
+    B: T.Buffer[(4, 4), "float32"],
+):
+    with T.block('compute'):
+        for io, ii in T.grid(4, 4):
+            if 0 <= 4*io + ii < 14:
+                B[io, ii] = 0.0
+                for f in T.serial(3):
+                    B[io, ii] = B[io, ii] + A[io + (ii + f) // 4, (ii + f) % 4]
+
+        for io,ii in T.grid(4,4):
+            if io==3 and ii>=2:
+                B[io,ii] = 0.0
+```
+
+
+We'd like to remove the conditional `if 0 <= 4*io + ii < 14` in the
+compute loop.  In order to do so, we need to prove that the body of
+the conditional is a no-op in the case where the conditional is false.
+
+Using the [updated `DomainTouched`
+utility](#enhancement-remove-no-op), this else-block would be a no-op.
+It is a write to `B[io,ii]` predicated on `4*io+ii >= 14`, followed by
+a write to `B[io,ii]` predicated on `io==3 and ii>=2`, without a read
+in between.  Since these predicates are equivalent, the first write is
+a no-op.
+
+```python
+# sched.remove_branching_through_overcompute(block='compute')
```

Comment:
Does this only apply to outputs?  I think we should per-buffer directive that indicates that out-of-bounds access is allowed. The only thing in question is how to determine/specify that out-of-bounds reads from inputs is ok.  The user can add padding -INF to inputs to maxpool, but how does the maxpool compute know that it can use the out-of-bounds values?

As to whether to actually utilize this should probably be left to the compiler.  Auto-scheduling should not be a replacement for compiler optimizations.